### PR TITLE
[android] update gradle and implement signing using gradle

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Update gradle to 7.0.2 and migrate jarsigner usage to have gradle do the align+sign automatically. Allows us to use the newer signing schemes (V2,v3. Dont intend for v4 at this stage)

## Motivation and context
Update gradle, and build a complete usable apk on macos host.

This implements some work @lrusak has previously done without the large sweeping cmake changes originally intended.

## How has this been tested?
Build macos/linux host, aarch64 android target
Jenkins test run https://jenkins.kodi.tv/job/Android-ARM64/20404/

with this PR
```
Library/Android/sdk/build-tools/31.0.0/apksigner verify --verbose --print-certs kodi-20211127-14e75479-macosarm_android-arm64-v8a.apk
Verifies
Verified using v1 scheme (JAR signing): true
Verified using v2 scheme (APK Signature Scheme v2): true
Verified using v3 scheme (APK Signature Scheme v3): true
Verified using v4 scheme (APK Signature Scheme v4): false
Verified for SourceStamp: false
Number of signers: 1
Signer #1 certificate DN: CN=XBMC Foundation, OU=Android platform, O=XBMC Foundation, L=US, ST=US, C=us
```

without
```
/Library/Android/sdk/build-tools/31.0.0/apksigner verify --verbose --print-certs kodi-20211127-abd2c059-PR20552-merge-arm64-v8a.apk
Verifies
Verified using v1 scheme (JAR signing): true
Verified using v2 scheme (APK Signature Scheme v2): false
Verified using v3 scheme (APK Signature Scheme v3): false
Verified using v4 scheme (APK Signature Scheme v4): false
Verified for SourceStamp: false
Number of signers: 1
Signer #1 certificate DN: CN=XBMC Foundation, OU=Android platform, O=XBMC Foundation, L=US, ST=US, C=us
```

## What is the effect on users?
N/A.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: tools/android/packaging/gradle.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+android.useAndroidX=true
+org.gradle.jvmargs=-Xmx4608m
```
Comment (lrusak) [inline tools/android/packaging/gradle.properties:2]: This is maximum memory usage right? It seems like a strange value?

File: tools/android/packaging/xbmc/build.gradle.in [hunk 2]
```diff
@@ -1,26 +1,35 @@
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 29
-    buildToolsVersion "29.0.3"
+    compileSdkVersion 31
```
Comment (lrusak) [inline tools/android/packaging/xbmc/build.gradle.in:4]: This seems unrelated to this PR and likely has other impacts. Is there a reason you need to do this here?

File: tools/android/packaging/xbmc/build.gradle.in [hunk 3]
```diff
@@ -1,26 +1,35 @@
...
-            signingConfig null
+    signingConfigs {
+        release {
+            keyAlias System.getenv("KODI_ANDROID_KEY_ALIAS")
+            keyPassword System.getenv("KODI_ANDROID_KEY_PASSWORD")
+            storeFile file(System.getenv("KODI_ANDROID_STORE_FILE"))
+            storePassword System.getenv("KODI_ANDROID_STORE_PASSWORD")
+            enableV1Signing true
+            enableV2Signing true
+            enableV3Signing true
```
Comment (lrusak) [inline tools/android/packaging/xbmc/build.gradle.in:20]: Can you explain what this is and why we want it?

**Review by lrusak — COMMENTED** [review]
Just left a few comments and questions. The rest looks good and is a good bridge between current master and the cmake PR.

File: tools/android/packaging/xbmc/build.gradle.in (hunk 3 above)
Comment (fuzzard) [reply tools/android/packaging/xbmc/build.gradle.in:20]: Android has incremental signing schemes that bring in new features/securities. jarsigner is only generic and not android specific, so it only does v1 signing. The transition to apksigner (either explicitly or gradle uses internally) brings the opportunity to sign using the newer schemes with no change to the end user.

V2 scheme brings in the more "secure" signing. v3's major feature is about key rotation. Only good things really, even if we dont use them (ie key rotation).

v4 scheme starts to bring in more changes like differentiation between upload signing and release signing. However it produces an idsig file. Until we need to go with v4, i dont think dealing with that is worth the hassle right now.

Comment (fuzzard) [reply tools/android/packaging/xbmc/build.gradle.in:20]: I should mention also, if a device runtime doesnt support a newer signing mechanism, it just falls back to the lower level.

File: tools/android/packaging/gradle.properties (hunk 1 above)
Comment (wsnipex) [reply tools/android/packaging/gradle.properties]: I guess he meant that 4608 is an unusual value, did you intend 4096m?

Comment (fuzzard) [reply tools/android/packaging/gradle.properties]: Didn't really matter tbh, got me to run some further tests, down to smallest simple number (1024)

File: tools/android/packaging/xbmc/build.gradle.in (hunk 3 above)
Comment (fuzzard) [reply tools/android/packaging/xbmc/build.gradle.in:20]: just reread this, and possibly you are asking about the syntax change. Somewhere i think in gradle 4 the syntax for signing changed from the one you had implemented into this. v3 scheme was done after the transition, so there is no v3SigningEnabled, hence the transition to the new style to support all currently available schems

**Comment by lrusak** [comment]
Has anyone tested a test build with this signing change?

**Comment by fuzzard** [comment]
I have, but anyone else feel free to give feedback

**Comment by Maven85** [comment]
I've tested this PR with my builds and works as expected.

However, I have to enter the value "android" for "KODI_ANDROID_KEY_PASSWORD" because I followed these instructions for the sign key: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md#33-create-a-key-to-sign-debug-apks

**Comment by fuzzard** [comment]
Yeah, its a bit odd as docs say to create the key with a password, but then the existing signing doesnt use it if its not set externally as well. I just leave behaviour as is, however previously it was a separate apk-sign target, now its a single apk target that signs as part of that build process.

I can set it to android as per docs as a default if people want it that way, not fussed really

**Comment by wsnipex** [comment]
yes please, docs should be consistent with behavior

**Comment by fuzzard** [comment]
Any further issues? Good to merge from my end

**Comment by Maven85** [comment]
No problems here.
